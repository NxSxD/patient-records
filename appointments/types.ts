export interface AppointmentLocation {
  lat: number;
  lng: number;
}

export interface Appointment {
  id?: string;
  doctor: string;
  appointment_time: string;
  location: AppointmentLocation;
}