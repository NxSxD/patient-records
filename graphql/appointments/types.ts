export interface AppointmentLocation {
  lat: number;
  lng: number;
}

export interface Appointment {
  appointment_time: string;
  doctor: string;
  location: AppointmentLocation;
}
