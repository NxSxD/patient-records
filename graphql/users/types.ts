export interface User {
  username: string;
  email: string;
  password: string;
}

export interface Credentials {
  email: string;
  password: string;
}
