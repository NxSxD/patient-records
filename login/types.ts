export interface AuthResponse {
  login: {
    token: string;
  };
}

export interface LoginVariables {
  email: string;
  password: string;
}
