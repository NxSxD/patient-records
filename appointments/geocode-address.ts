import { AppointmentLocation } from "./types";

export async function geocodeAddress(address: string): Promise<AppointmentLocation | null>{
  let location;

  try {
    const geocodingURL = new URL(
      "https://maps.googleapis.com/maps/api/geocode/json"
    );
    const params = {
      key: "AIzaSyDWO-NcYK0BFVe4MfRZ1iDumC8Q535jANo",
      address,
    };

    geocodingURL.search = new URLSearchParams(params).toString();

    const response = await fetch(geocodingURL.toString(), {
      method: "GET",
    });

    if (response.headers.get("Content-Type").includes("application/json")) {
      const data = await response.json();
      const firstResult = data.results[0];
      location = {
        lat: firstResult.geometry.location.lat,
        lng: firstResult.geometry.location.lng,
      };
    }
  } catch (e) {
    console.log("E: ", e);
    return null;
  }

  return location;
}
