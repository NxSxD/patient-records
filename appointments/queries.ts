import { gql } from "@apollo/react-hooks";
import { Appointment } from "./types";

// Get last 3 appointments
export interface GetAppointmentsVariables {}

export interface GetAppointmentsData {
  appointments: Appointment[]
}

export const GET_APPOINTMENTS = gql`
  query getAppointments {
    appointments(
      limit: 3,
      orderBy: "appointment_time"
    ) {
      id
      doctor
      appointment_time
      location {
        lat
        lng
      }
    }
  }
`;

// Get appointment by ID
export interface GetAppointmentVariables {}

export interface GetAppointmentData {
  appointment: Appointment;
}

export const GET_APPOINTMENT = gql`
  query getAppointment($id: ID!) {
    appointment(id: $id) {
      id
      doctor
      appointment_time
      location {
        lat
        lng
      }
    }
  }
`;

// Add Appointment
export interface AddAppointmentVariables {
  appointment: Omit<Appointment, 'id'>;
}

export interface AddAppointmentData {
  addAppointment: Appointment;
}

export const ADD_APPOINTMENT = gql`
  mutation addAppointment($appointment: NewAppointmentPayload) {
    addAppointment(appointment: $appointment) {
      id
      doctor
      appointment_time
      location {
        lat
        lng
      }
    }
  }
`;