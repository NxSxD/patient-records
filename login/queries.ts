import { gql } from "@apollo/react-hooks";

export interface AuthResponse {
  login: {
    token: string;
  };
}

export interface LoginVariables {
  email: string;
  password: string;
}

export const LOGIN = gql`
  mutation Login($email: String!, $password: String!) {
    login(credentials: {
      email: $email,
      password: $password
    }) {
      token
    }
  }
`;
