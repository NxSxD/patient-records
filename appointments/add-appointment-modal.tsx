import React, { useEffect } from "react";
import styled from "styled-components";
import {
  Mask,
  ModalCard,
  ModalHeader,
  ButtonPrimary,
  Form,
  Spinner,
} from "../primitives";
import { Formik, FormikHelpers } from "formik";
import { Field } from "../components";
import { Appointment } from "./types";
import dayjs from "dayjs";
import { geocodeAddress } from "./geocode-address";
import {
  ADD_APPOINTMENT,
  AddAppointmentData,
  AddAppointmentVariables,
} from "./queries";
import { useMutation } from "@apollo/react-hooks";

interface AddAppointmentModalProps extends React.RefAttributes<HTMLDivElement> {
  onAppointmentAdded: (data: Appointment) => void;
}

interface FormValues {
  doctor: string;
  date: string;
  time: string;
  location: string;
}

export const AddAppointmentModal = React.forwardRef<
  HTMLDivElement,
  AddAppointmentModalProps
>(function addAppointmentModal({ onAppointmentAdded }, ref) {
  const values: FormValues = {
    doctor: "",
    date: "",
    time: "",
    location: "",
  };

  const [addAppointment, { data, loading }] = useMutation<
    AddAppointmentData,
    AddAppointmentVariables
  >(ADD_APPOINTMENT);

  useEffect(() => {
    if (!data || !data.addAppointment) return;
    if (data.addAppointment.id) onAppointmentAdded(data.addAppointment);
  }, [data]);

  const onSubmit = async (
    values: FormValues,
    helpers: FormikHelpers<FormValues>
  ) => {
    if (!values.doctor) {
      helpers.setErrors({ doctor: "Doctor's name is required." });
      return;
    }

    if (!values.date) {
      helpers.setErrors({ date: "Appointment date is required." });
      return;
    }

    if (!values.time) {
      helpers.setErrors({ time: "Appointment time is required." });
      return;
    }

    if (!values.location) {
      helpers.setErrors({ location: "Appointment location is required." });
      return;
    }

    const appointmentLocation = await geocodeAddress(values.location);
    if (!appointmentLocation) {
      helpers.setErrors({ location: "Invalid address provided." });
      return;
    }

    const [hour, minute] = values.time.split(":");
    const appointment_time = dayjs(values.date)
      .hour(parseInt(hour))
      .minute(parseInt(minute))
      .toISOString();
    const appointment: Appointment = {
      doctor: values.doctor,
      appointment_time,
      location: appointmentLocation,
    };

    addAppointment({
      variables: {
        appointment,
      },
    });

    helpers.setSubmitting(false);
  };

  return (
    <PageMask>
      <ModalCard ref={ref}>
        <ModalHeader>New Appointment</ModalHeader>
        <FormContainer>
          <Formik initialValues={values} onSubmit={onSubmit}>
            {({ handleSubmit, isSubmitting }) => (
              <Form onSubmit={handleSubmit}>
                <Field label="Doctor" name="doctor" type="text" />
                <Field label="Date" name="date" type="date" />
                <Field label="Time" name="time" type="time" />
                <Field label="Location" name="location" type="text" />
                <ButtonPrimary type="submit" disabled={isSubmitting || loading}>
                  {isSubmitting ? (
                    <Spinner aria-label="Loading..." />
                  ) : (
                    "Add Appointment"
                  )}
                </ButtonPrimary>
              </Form>
            )}
          </Formik>
        </FormContainer>
      </ModalCard>
    </PageMask>
  );
});

const PageMask = styled(Mask)`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  justify-content: center;
  align-items: flex-end;

  @media screen and (min-width: 720px) {
    align-items: center;
  }
`;

const FormContainer = styled.div`
  flex: 1;
  padding: 1rem 1rem;
  display: flex;
  flex-direction: column;
  justify-content: flex-end;
`;
