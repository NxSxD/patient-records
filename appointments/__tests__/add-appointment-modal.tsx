import React from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { AddAppointmentModal } from "../add-appointment-modal";
import { ADD_APPOINTMENT } from "../queries";
import { Appointment } from "../types";
// @ts-ignore
import { MockedProvider } from "@apollo/client/testing";
import { geocodeAddress } from "../geocode-address";

jest.mock("../geocode-address", () => ({
  geocodeAddress: jest.fn().mockResolvedValue({ lat: 25, lng: 25 }),
}));

const date = "2020-08-28";
const time = "13:30";
const address = "459 Brandon Blvd, Brandon, Fl";
const submitButtonText = "Add Appointment";
const newAppointment: Appointment = {
  id: "123",
  doctor: "Doctor 1",
  appointment_time: "2020-08-28T17:30:00.000Z",
  location: {
    lat: 25,
    lng: 25,
  },
};

const mocks = [
  {
    request: {
      query: ADD_APPOINTMENT,
      variables: {
        appointment: {
          doctor: newAppointment.doctor,
          appointment_time: newAppointment.appointment_time,
          location: newAppointment.location,
        },
      },
    },
    result: {
      data: { addAppointment: newAppointment },
    },
  },
];

describe("AddAppointmentModal", () => {
  const onAppointmentAdded = jest.fn();

  beforeEach(() => {
    onAppointmentAdded.mockReset();
  });

  it("should render according to snapshot", () => {
    const { container } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <AddAppointmentModal onAppointmentAdded={onAppointmentAdded} />
      </MockedProvider>
    );

    expect(container).toMatchSnapshot();
  });

  it("should invoke onAddAppointment", async () => {
    const { getByText } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <AddAppointmentModal onAppointmentAdded={onAppointmentAdded} />
      </MockedProvider>
    );

    fireEvent.input(screen.getByLabelText(/Doctor-input/), {
      target: { value: newAppointment.doctor },
    });

    fireEvent.change(screen.getByLabelText(/Date-input/), {
      target: { value: date },
    });

    fireEvent.input(screen.getByLabelText(/Time-input/), {
      target: { value: time },
    });

    fireEvent.input(screen.getByLabelText(/Location-input/), {
      target: { value: address },
    });

    fireEvent.click(screen.getByText(submitButtonText));

    await waitFor(() => {
      expect(getByText(submitButtonText)).toBeInTheDocument();
    });

    expect(onAppointmentAdded).toHaveBeenCalled();
  });

  describe("Validate required fields", () => {
    test.each([
      [
        "Doctor's name is required.",
        [/Date-input/, /Time-input/, /Location-input/],
        [date, time, address],
      ],
      [
        "Appointment date is required.",
        [/Doctor-input/, /Time-input/, /Location-input/],
        [newAppointment.doctor, time, address],
      ],
      [
        "Appointment time is required.",
        [/Doctor-input/, /Date-input/, /Location-input/],
        [newAppointment.doctor, date, address],
      ],
      [
        "Appointment location is required.",
        [/Doctor-input/, /Date-input/, /Time-input/],
        [newAppointment.doctor, date, time],
      ],
    ])("should display error - %s", async (expected, inputs, values) => {
      const { getByText } = render(
        <MockedProvider mocks={mocks} addTypename={false}>
          <AddAppointmentModal onAppointmentAdded={onAppointmentAdded} />
        </MockedProvider>
      );

      for (let i = 0; i < inputs.length; i++) {
        fireEvent.change(screen.getByLabelText(inputs[i]), {
          target: { value: values[i] },
        });
      }

      fireEvent.click(screen.getByText(submitButtonText));

      await waitFor(() => {
        expect(getByText(submitButtonText)).toBeInTheDocument();
      });

      expect(onAppointmentAdded).not.toHaveBeenCalled();
      expect(screen.getByRole("alert")).toHaveTextContent(expected);
    });

    describe("Inavlid gecode", () => {
      beforeEach(() => {
        // @ts-ignore
        geocodeAddress.mockReset().mockResolvedValue(null);
      });
      it("should display error: Invalid address provided.", async () => {
        const { getByText, getByLabelText } = render(
          <MockedProvider mocks={mocks} addTypename={false}>
            <AddAppointmentModal onAppointmentAdded={onAppointmentAdded} />
          </MockedProvider>
        );

        fireEvent.input(getByLabelText(/Doctor-input/), {
          target: { value: newAppointment.doctor },
        });

        fireEvent.change(getByLabelText(/Date-input/), {
          target: { value: date },
        });

        fireEvent.input(getByLabelText(/Time-input/), {
          target: { value: time },
        });

        fireEvent.input(getByLabelText(/Location-input/), {
          target: { value: "some random value" },
        });

        fireEvent.click(getByText(submitButtonText));

        await waitFor(() => {
          expect(getByText(submitButtonText)).toBeInTheDocument();
        });

        expect(onAppointmentAdded).not.toHaveBeenCalled();
        expect(screen.getByRole("alert")).toHaveTextContent(
          "Invalid address provided."
        );
      });
    });
  });
});
