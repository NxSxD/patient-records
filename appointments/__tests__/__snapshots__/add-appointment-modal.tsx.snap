// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddAppointmentModal should render according to snapshot 1`] = `
.c8 {
  height: 45px;
  padding: 0 1rem;
  font-size: 1.2rem;
  border: none;
  border-radius: 4px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-transition: background-color 250ms;
  transition: background-color 250ms;
  min-width: 120px;
  background-color: var(--primary-color);
  color: #fff;
}

.c8:hover {
  cursor: pointer;
  background-color: var(--primary-alt-color);
}

.c8:focus {
  outline: none;
}

.c1 {
  background: #fff;
  border-radius: 4px;
  min-height: 450px;
  width: 80%;
  max-width: 800px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c2 {
  height: 45px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0 1rem;
  font-size: 1.2rem;
  color: #999;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  margin-bottom: 2rem;
}

.c6 {
  color: #999;
  font-size: 1.1rem;
  margin-bottom: 5px;
}

.c7 {
  border: 1px solid #ccc;
  border-radius: 4px;
  background: #fff;
  font-size: 1.2rem;
  color: #666;
  padding-left: 12px;
  height: 45px;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 {
  background: linear-gradient(transparent,rgba(0,0,0,0.7));
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
}

.c3 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  padding: 1rem 1rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

@media screen and (min-width:720px) {
  .c0 {
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
  }
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        New Appointment
      </div>
      <div
        class="c3"
      >
        <form
          class="c4"
        >
          <div
            class="c5"
          >
            <span
              class="c6"
            >
              Doctor
            </span>
            <input
              aria-label="Doctor-input"
              class="c7"
              name="doctor"
              type="text"
              value=""
            />
          </div>
          <div
            class="c5"
          >
            <span
              class="c6"
            >
              Date
            </span>
            <input
              aria-label="Date-input"
              class="c7"
              name="date"
              type="date"
              value=""
            />
          </div>
          <div
            class="c5"
          >
            <span
              class="c6"
            >
              Time
            </span>
            <input
              aria-label="Time-input"
              class="c7"
              name="time"
              type="time"
              value=""
            />
          </div>
          <div
            class="c5"
          >
            <span
              class="c6"
            >
              Location
            </span>
            <input
              aria-label="Location-input"
              class="c7"
              name="location"
              type="text"
              value=""
            />
          </div>
          <button
            class="c8"
            type="submit"
          >
            Add Appointment
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;
